src/web/lib/braze/checkBrazeDependencies.ts

```diff
@@ -1,63 +1,96 @@
-import {
-	checkDependencies,
-	DependencyConfig,
-	DependencyResult,
-} from '@root/src/web/lib/dependencyChecker';
 import { getBrazeUuid } from '@root/src/web/lib/getBrazeUuid';
 import { hasRequiredConsents } from './hasRequiredConsents';
 import { hideSupportMessaging } from './hideSupportMessaging';
 
-const getCompleteDependencies = async (
-	isSignedIn: boolean,
-	idApiUrl: string,
-): Promise<Array<DependencyConfig>> => {
-	return [
-		{
-			name: 'brazeSwitch',
-			dependency: Promise.resolve(
-				window.guardian.config.switches.brazeSwitch,
-			),
-		},
-		{
-			name: 'apiKey',
-			dependency: Promise.resolve(
-				window.guardian.config.page.brazeApiKey,
-			),
-		},
-		{
-			name: 'consent',
-			dependency: hasRequiredConsents(),
-		},
-		{
-			name: 'isNotPaidContent',
-			dependency: Promise.resolve(
-				!window.guardian.config.page.isPaidContent,
-			),
-		},
-		{
-			name: 'brazeUuid',
-			dependency: isSignedIn
-				? getBrazeUuid(idApiUrl)
-				: Promise.resolve(null),
-		},
-		{
-			name: 'userIsGuSupporter',
-			// Currently all active web canvases in Braze target existing supporters,
-			// subscribers or otherwise those with a Guardian product. We can use the
-			// value of `shouldHideSupportMessaging` to identify these users, limiting
-			// the number of requests we need to initialise Braze on the page:
-			dependency: Promise.resolve(hideSupportMessaging()),
-		},
-	];
+type SuccessResult = {
+	isSuccessful: true;
+	data: ResultData;
+};
+
+type FailureResult = {
+	isSuccessful: false;
+	failureField: string;
+	failureData: any;
+	data: ResultData;
 };
 
+type DependencyResult = SuccessResult | FailureResult;
+
+type ResultData = { [key: string]: any };
+
+const buildFailureResponse = (name: string, value: any, data: ResultData) => ({
+	isSuccessful: false,
+	failureField: name,
+	failureData: value,
+	data,
+});
+
 const checkBrazeDependencies = async (
 	isSignedIn: boolean,
 	idApiUrl: string,
 ): Promise<DependencyResult> => {
-	return checkDependencies(
-		await getCompleteDependencies(isSignedIn, idApiUrl),
-	);
+	const data: ResultData = {};
+
+	{
+		const name = 'brazeSwitch';
+		const value = window.guardian.config.switches.brazeSwitch;
+		if (!value) {
+			return buildFailureResponse(name, value, data);
+		}
+		data[name] = value;
+	}
```

Comment:
The structure I has in mind was more on the lines of:

```
const brazeSwitch = new Promise((resolve, reject) => {
    const value = window.guardian.config.switches.brazeSwitch;
    if (!value) {
        reject(buildFailureResponse(name, value, data))
    }
    resolve(value)
});

[...]

return [brazeSwitch, ...]
```

That means when you call `checkBrazeDependencies` you can do something like:

```
const got = await Promise.all[checkBrazeDependencies(isSignedIn, idApiUrl)}
    .then(results => {
        ...
    })
    .catch(err => {
        ...
    });
```

Also in thise case it might be worth renaming `checkBrazeDependencies` to something like `arrayBrazeChecks` (not fully convinced by the my suggestion either)

The example I based on a post I found really useful https://www.javascripttutorial.net/es6/javascript-promise-all/
Thoughts on this?

Reply:
Thanks for the detail! I considered `Promise.all` before, but ended up discounting it as the behaviour is _slightly_ different to what we have now. Currently the failure (if any) will be for the first item in the ordered list of dependencies. Using `Promise.all` changes that slightly to be the first item which fails chronologically, which _might_ not be the same. Especially if some operations are longer running than others (for example an API call). I'm now wondering how relevant or important that is though.

The other piece I'm not sure about with `Promise.all` is that in the event that one of the checks fails, we want to be able to access the data from any other successful checks. I don't think `Promise.all` lets you access that data (which is why we were reducing a list of promises in the `dependencyChecker`).

I definitely agree that it would be nice to be using common JavaScript APIs and building blocks here. I'll try some things out and report back.

Reply:
I see, maybe it might be worth looking at using `Promise.allSettled()` https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled

From what I gather that might be a better than `Promise.all ` for what you are hoping to achieve?

Reply:
Hi @liywjl I had a play with `Promise.allSettled` but wasn't super happy with the result (it turns out it isn't supported in the version of Node we're on which adds some additional friction). I pushed up a few more commits, including f160afb, which is kind of a compromise approach where I iterate over the conditions rather than pass them to `Promise.allSetted`. I think it could be a nice happy medium but interested in your take on it.

I left the `allSetted` approach in there but commented out so you can see how it looked. The part where we iterate the results to build up the final `DependenciesResult` object wasn't very nice.

Reply:
Hi @liywjl I've had a bit more of a play around with refactoring this but didn't get to a point where I was happier with the result than what's there now. I've removed the commented `Promise.allSettled` experimental code and have left a comment on using `Promise.allSettled` in future when it's more reliably available. What are your thoughts on me merging this as it is?